### Assignment 4 (12th September 2022, Monday)


#### Initializing Connection


In [1]:
import mysql.connector as sqltor

connection = sqltor.connect(host="localhost", user="root", password="mysqlpasswd")
cursor = connection.cursor()
cursor.execute("CREATE DATABASE IF NOT EXISTS week4")
cursor.execute("USE week4")


#### Adding Helper functions


In [2]:
def disp():
    print(*[line for line in cursor], sep="\n")


def run(query: str):
    cursor.execute(query)


#### Testing Tables and Datasets


In [3]:
run("SHOW DATABASES")
disp()
run("SHOW TABLES")
disp()
# exit()


('information_schema',)
('mysql',)
('performance_schema',)
('sys',)
('week1',)
('week2',)
('week3',)
('week3i',)
('week4',)



| Database           |
| ------------------ |
| information_schema |
| mysql              |
| performance_schema |
| sys                |
| week1              |
| week2              |
| week3              |
| week3i             |
| week4              |


#### Queries


##### 1. Show the use of upper and lower function.


In [4]:
run("SELECT UPPER('hello world') AS UPPER")
disp()
run("SELECT LOWER('HELLO WORLD') AS LOWER")
disp()


('HELLO WORLD',)
('hello world',)


TABLE PREVIEW


| UPPER       |
| :---------- |
| HELLO WORLD |

| LOWER       |
| :---------- |
| hello world |


##### 2. Show the use of concat, instr and length function.


In [5]:
run("SELECT CONCAT('HELLO', 'WORLD') AS CONCATENATE")
disp()
run("SELECT INSTR('Pritam A40', 'A') AS 'FIND STRING'")
disp()


('HELLOWORLD',)
(5,)


TABLE PREVIEW


| CONCATENATE |
| :---------- |
| HELLOWORLD  |

| FIND STRING |
| :---------- |
| 5           |


##### 3. Show the use of the following functions on numeric values:

    i. Sqrt()
    ii. Power()
    iii. Ceil()
    iv. Substr()
    v. Max()
    vi. Min()
    vii. Round()
    viii. Avg()
    ix. Count()
    x. Exp()
    xi. Mod()


PRE REQUISITE


In [6]:
run("CREATE TABLE IF NOT EXISTS INTEGERS(NUMBER INT)")
run("TRUNCATE INTEGERS")
run("INSERT INTO INTEGERS VALUES (1), (2), (3), (4), (5)")
connection.commit()


QUERIES


In [ ]:
run("SELECT SQRT(1024) AS SQRT")
disp()
run("SELECT POWER(4, 2) AS POWER")
disp()
run("SELECT CEIL(4.9) AS CEIL")
disp()
run("SELECT SUBSTR('PRITAM KUNDU', 1, 6) AS SUBSTRING")
disp()
run("SELECT MAX(NUMBER) AS MAXIMUM FROM INTEGERS")
disp()
run("SELECT MIN(NUMBER) AS MINIMUM FROM INTEGERS")
disp()
run("SELECT ROUND(9.98) AS ROUND")
disp()
run("SELECT AVG(NUMBER) AS AVERAGE FROM INTEGERS")
disp()
run("SELECT COUNT(*) AS `NUMBER OF INTEGERS` FROM INTEGERS")
disp()
run("SELECT EXP(3) AS EXPONENT")
disp()
run("SELECT MOD(7, 3) AS MODULUS")
disp()


TABLE PREVIEW


| SQRT |
| :--- |
| 32   |

| POWER |
| :---- |
| 16    |


##### 4. Solve the following queries

    1. Find the ceiling and floor value of 14.887
    2. Find out the round off of 17.49989
    3. Calculate 8<sup>7</sup>


In [43]:
run("SELECT CEIL(14.887), FLOOR(14.887)")
disp()
run("SELECT ROUND(17.49989)")
disp()
run("SELECT POW(8, 7)")
disp()


(15, 14)
(Decimal('17'),)
(2097152.0,)


TABLE PREVIEW


| CEIL(14.887) | FLOOR(14.887) |
| ------------ | ------------- |
| 15           | 14            |

| ROUND(17.49989) |
| --------------- |
| 17              |

| POW(8, 7) |
| --------- |
| 2097152   |


##### 5. Show the current date


In [45]:
run("SELECT CURDATE()")
disp()


(datetime.date(2022, 9, 13),)


TABLE PREVIEW


| CURDATE()  |
| ---------- |
| 2022-10-16 |


##### 6. Find the total experience of the employees in weeks who works in Sales department


In [ ]:
run(
    """
    SELECT SUM(Experience) FROM Employees 
    WHERE Dept = 'Sales' GROUP BY WEEK
"""
)
disp()


TABLE PREVIEW


| SUM(Experience) |
| --------------: |
|             100 |
|              20 |
|              40 |


##### 7. Display the use of the following functions on date

    1. Months_between
    2. Add_months
    3. Next_day
    4. last_day
    5. Round
    6. Trunc
    7. To_char


In [66]:
run("SELECT TIMESTAMPDIFF(month, '2021-05-18', '2021-07-29') AS MONTH_DIFF")
disp()
run("SELECT DATE_ADD('2002-03-15', INTERVAL 1 MONTH) AS MONTH_ADD")
disp()
run("SELECT DATE_ADD(NOW(), INTERVAL 1 DAY)")
disp()
run("SELECT DATE_SUB(NOW(), INTERVAL 1 DAY)")
disp()
# 5
# 6
run("SELECT DATE_FORMAT(NOW(), '%D-%M-%Y')")
disp()


(2,)
('2002-04-15',)
(datetime.datetime(2022, 9, 14, 15, 19, 14),)
(datetime.datetime(2022, 9, 12, 15, 19, 14),)
('13th-September-2022',)


TABLE PREVIEW


| MONTH_DIFF |
| ---------- |
| 2          |

| MONTH_ADD  |
| ---------- |
| 2002-04-15 |


##### 8. Show the employee details with a revised salary. The salary is incremented in the following way:

    1. 10% for sales department
    2. 20% for marketing department
    3. No increment for others


In [ ]:
run(
    """

"""
)
disp()


TABLE PREVIEW


##### 9. Determine the tax for each employee in production department based on the monthly salary. The tax rate are as per the following data:

| Monthly Salary Range | Rate |
| :------------------- | :--- |
| 0-19,999             | 0%   |
| 20,000-39,999        | 9%   |
| 40,000-59,999        | 20%  |
| 60,000-79,999        | 30%  |
| 80,000 or more       | 45%  |


TABLE PREVIEW


#### 10. Find the Cartesian product between Employee and Department table.


In [ ]:
run("SELECT * FROM employee JOIN department")
disp()


TABLE PREVIEW


| Emp_id | F_Name  | L_Name | Job_Type   | Salary | Comission | Dept       | Manager_id | DOJ        | d_name     | d_loc     | HOD_id |
| ------ | ------- | ------ | ---------- | ------ | --------- | ---------- | ---------- | ---------- | ---------- | --------- | ------ |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | R & D      | Marketing | 8      |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | Marketing  | Kol       | 2      |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | Production | Kol       | 1      |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | Accounts   | Delhi     | 6      |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | Sales      | Kol       | 4      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | R & D      | Marketing | 8      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | Marketing  | Kol       | 2      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | Production | Kol       | 1      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | Accounts   | Delhi     | 6      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | Sales      | Kol       | 4      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | R & D      | Marketing | 8      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | Marketing  | Kol       | 2      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | Production | Kol       | 1      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | Accounts   | Delhi     | 6      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | Sales      | Kol       | 4      |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | R & D      | Marketing | 8      |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | Marketing  | Kol       | 2      |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | Production | Kol       | 1      |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | Accounts   | Delhi     | 6      |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | Sales      | Kol       | 4      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | R & D      | Marketing | 8      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | Marketing  | Kol       | 2      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | Production | Kol       | 1      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | Accounts   | Delhi     | 6      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | Sales      | Kol       | 4      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | R & D      | Marketing | 8      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | Marketing  | Kol       | 2      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | Production | Kol       | 1      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | Accounts   | Delhi     | 6      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | Sales      | Kol       | 4      |


#### 11. Show the employee names and the respective department location


In [ ]:
run(
    """SELECT 
           CONCAT(f_name,' ' ,l_name) AS NAME, 
           d_loc FROM employee, department 
       WHERE Dept = d_name
    """
)
disp()


TABLE PREVIEW


| NAME          | d_loc |
| ------------- | ----- |
| Arun Khan     | Kol   |
| Barun Kumar   | Kol   |
| Chitra Kapoor | Kol   |
| Dheeraj Mitra | Kol   |
| Emma Dutt     | Kol   |
| Floki Dutt    | Delhi |


#### 12. Give an example of the following joins considering employee and department tables.

    1. Natural join
    2. Inner join
    3. Left outer join
    4. Right outer join
    5. Full outer join


In [ ]:
run(
    "SELECT * FROM Employee NATURAL JOIN (SELECT d_loc, HOD_id, d_name as Dept FROM Department) AS _"
)
disp()


TABLE PREVIEW


| Dept       | Emp_id | F_Name  | L_Name | Job_Type   | Salary | Comission | Manager_id | DOJ        | d_loc | HOD_id |
| ---------- | ------ | ------- | ------ | ---------- | ------ | --------- | ---------- | ---------- | ----- | ------ |
| Production | 1      | Arun    | Khan   | Manager    | 90000  | NULL      | NULL       | 1998-01-04 | Kol   | 1      |
| Marketing  | 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | NULL       | 1998-02-09 | Kol   | 2      |
| Production | 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | 1          | 1998-01-08 | Kol   | 1      |
| Sales      | 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | 4          | 2001-12-27 | Kol   | 4      |
| Production | 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | 1          | 2002-03-20 | Kol   | 1      |
| Accounts   | 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | NULL       | 2000-07-16 | Delhi | 6      |


In [ ]:
run("SELECT * FROM Employee INNER JOIN Department ON Dept=D_name")
disp()


| Emp_id | F_Name  | L_Name | Job_Type   | Salary | Comission | Dept       | Manager_id | DOJ        | d_name     | d_loc | HOD_id |
| ------ | ------- | ------ | ---------- | ------ | --------- | ---------- | ---------- | ---------- | ---------- | ----- | ------ |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | Production | Kol   | 1      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | Marketing  | Kol   | 2      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | Production | Kol   | 1      |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | Sales      | Kol   | 4      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | Production | Kol   | 1      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | Accounts   | Delhi | 6      |


In [ ]:
run("SELECT * FROM Employee LEFT OUTER JOIN Department ON Dept=D_name")
disp()


| Emp_id | F_Name  | L_Name | Job_Type   | Salary | Comission | Dept       | Manager_id | DOJ        | d_name     | d_loc | HOD_id |
| ------ | ------- | ------ | ---------- | ------ | --------- | ---------- | ---------- | ---------- | ---------- | ----- | ------ |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | Production | Kol   | 1      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | Marketing  | Kol   | 2      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | Production | Kol   | 1      |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | Sales      | Kol   | 4      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | Production | Kol   | 1      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | Accounts   | Delhi | 6      |


In [ ]:
run("SELECT * FROM Employee RIGHT OUTER JOIN Department ON Dept=D_name")
disp()


| Emp_id | F_Name  | L_Name | Job_Type   | Salary | Comission | Dept       | Manager_id | DOJ        | d_name     | d_loc     | HOD_id |
| ------ | ------- | ------ | ---------- | ------ | --------- | ---------- | ---------- | ---------- | ---------- | --------- | ------ |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | Sales      | Kol       | 4      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | Accounts   | Delhi     | 6      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | Production | Kol       | 1      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | Production | Kol       | 1      |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | Production | Kol       | 1      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | Marketing  | Kol       | 2      |
| NULL   | NULL    | NULL   | NULL       | NULL   | NULL      | NULL       | NULL       | NULL       | R & D      | Marketing | 8      |


In [ ]:
run("SELECT * FROM Employee FULL JOIN Department ON Dept=D_name")
disp()


| Emp_id | F_Name  | L_Name | Job_Type   | Salary | Comission | Dept       | Manager_id | DOJ        | d_name     | d_loc | HOD_id |
| ------ | ------- | ------ | ---------- | ------ | --------- | ---------- | ---------- | ---------- | ---------- | ----- | ------ |
| 1      | Arun    | Khan   | Manager    | 90000  | NULL      | Production | NULL       | 1998-01-04 | Production | Kol   | 1      |
| 2      | Barun   | Kumar  | Manager    | 80000  | NULL      | Marketing  | NULL       | 1998-02-09 | Marketing  | Kol   | 2      |
| 3      | Chitra  | Kapoor | Engineer   | 60000  | NULL      | Production | 1          | 1998-01-08 | Production | Kol   | 1      |
| 4      | Dheeraj | Mitra  | Accountant | 75000  | NULL      | Sales      | 4          | 2001-12-27 | Sales      | Kol   | 4      |
| 5      | Emma    | Dutt   | Engineer   | 55000  | NULL      | Production | 1          | 2002-03-20 | Production | Kol   | 1      |
| 6      | Floki   | Dutt   | Accountant | 70000  | NULL      | Accounts   | NULL       | 2000-07-16 | Accounts   | Delhi | 6      |


#### 13. Write a query to find the addresses (location_id, street_address, city, state_province, country_name) of all the departments. USE NATURAL JOIN


In [ ]:
run(
    """CREATE TABLE locations(
           location_id INT(4) PRIMARY KEY,
           street_address VARCHAR(100),
           postal_code VARCHAR(20),
           city VARCHAR(30),
           state_province VARCHAR(30),
           country_id VARCHAR(2),
           FOREIGN KEY (country_id) REFERENCES countries(country_id)
      )
"""
)

run(
    """CREATE TABLE countries(
           country_id VARCHAR(2) PRIMARY KEY,
           country_name VARCHAR(20),
           region_id INT(1)
    )
"""
)

run(
    """
    INSERT INTO countries VALUES
        ('AR', 'Argentina', 2),
        ('AU', 'Australia', 3),
        ('BE', 'Belgium', 3),
        ('BR', 'Brazil', 3),
        ('CA', 'Canada', 3)
"""
)

run(
    """
    INSERT INTO locations VALUES
        (1000, '1297 Via Cola di Rie', 989, 'Roma', NULL, 'BE'),
        (1100, '93091 Calle della Te', 10934, 'Venice', NULL, 'BE')
"""
)


In [ ]:
run(
    """SELECT
           location_id, 
           street_address, 
           city, 
           state_province, 
           country_name 
      FROM countries NATURAL JOIN locations
"""
)


TABLE PREVIEW


| location_id | street_address       | city   | state_province | country_name |
| ----------- | -------------------- | ------ | -------------- | ------------ |
| 1000        | 1297 Via Cola di Rie | Roma   | NULL           | Belgium      |
| 1100        | 93091 Calle della Te | Venice | NULL           | Belgium      |


#### 14. Write a query to find the name (first name, last_name), department ID of all the employees.


In [ ]:
run(
    """SELECT
           CONCAT(firstName, lastName) AS Name,
           department_ID
      FROM employees NATURAL JOIN departments
"""
)


TABLE PREVIEW


| Name          | Department_ID |
| ------------- | ------------- |
| Steven King   | 10            |
| Nenna Kochhar | 20            |
